### ISSUE TITLE
iscsiadm does not mount devices if target does not contain LUN 0

### ISSUE THREAD

**Issue by morden2k** [open-iscsi/open-iscsi#175]
I have target with LUN's: `0, 20, 40`. All is work till I delete LUN `0` on target. Then iscsiadm does not mount any device on target. If I add LUN 0 again, then iscsiadm mounts devices again.

### Initial disks state:

```
$ ll /dev/disk/by-id/
total 0
drwxr-xr-x 2 root root  60 жов 16 11:17 ./
drwxr-xr-x 5 root root 100 жов  4 16:32 ../
lrwxrwxrwx 1 root root   9 жов  4 16:32 ata-VMware_Virtual_IDE_CDROM_Drive_10000000000000000001 -> ../../sr0
```
### Login to target with LUN 0:
```
$ sudo -s iscsiadm --mode node --portal 192.168.13.116:3260 --targetname iqn.2010-06.com.nutanix:volumegroup-2-0a51e6cb-1666-4feb-809c-8d38110f17c2 --login
$ sudo iscsiadm -m session -P 3
iSCSI Transport Class version 2.0-870
version 2.0-873
Target: iqn.2010-06.com.nutanix:volumegroup-2-0a51e6cb-1666-4feb-809c-8d38110f17c2 (non-flash)
	Current Portal: 192.168.13.116:3260,1
	Persistent Portal: 192.168.13.116:3260,1
		**********
		Interface:
		**********
		Iface Name: default
		Iface Transport: tcp
		Iface Initiatorname: iqn.1993-08.org.debian:01:5bb3c3d5f39
		Iface IPaddress: 192.168.12.227
		Iface HWaddress: <empty>
		Iface Netdev: <empty>
		SID: 363
		iSCSI Connection State: LOGGED IN
		iSCSI Session State: LOGGED_IN
		Internal iscsid Session State: NO CHANGE
		*********
		Timeouts:
		*********
		Recovery Timeout: 120
		Target Reset Timeout: 30
		LUN Reset Timeout: 30
		Abort Timeout: 15
		*****
		CHAP:
		*****
		username: <empty>
		password: ********
		username_in: <empty>
		password_in: ********
		************************
		Negotiated iSCSI params:
		************************
		HeaderDigest: None
		DataDigest: None
		MaxRecvDataSegmentLength: 262144
		MaxXmitDataSegmentLength: 1048576
		FirstBurstLength: 262144
		MaxBurstLength: 16776192
		ImmediateData: Yes
		InitialR2T: Yes
		MaxOutstandingR2T: 1
		************************
		Attached SCSI devices:
		************************
		Host Number: 365	State: running
		scsi365 Channel 00 Id 0 Lun: 0
			Attached scsi disk sdc		State: running
		scsi365 Channel 00 Id 0 Lun: 20
			Attached scsi disk sde		State: running
		scsi365 Channel 00 Id 0 Lun: 40
			Attached scsi disk sdd		State: running

			
$ ll /dev/disk/by-id/
total 0
drwxr-xr-x 2 root root 120 жов 16 11:25 ./
drwxr-xr-x 5 root root 100 жов  4 16:32 ../
lrwxrwxrwx 1 root root   9 жов  4 16:32 ata-VMware_Virtual_IDE_CDROM_Drive_10000000000000000001 -> ../../sr0
lrwxrwxrwx 1 root root   9 жов 16 11:25 scsi-1NUTANIX_NFS_2_0_2565_687fd312_fddb_40a7_a2b7_14c26a063d44 -> ../../sde
lrwxrwxrwx 1 root root   9 жов 16 11:25 scsi-1NUTANIX_NFS_2_0_2567_bb47733b_115a_4625_ba72_496d994129e9 -> ../../sdd
lrwxrwxrwx 1 root root   9 жов 16 11:25 scsi-1NUTANIX_NFS_2_0_2570_b6cb89fd_23dd_4444_8e9b_4448694f92b9 -> ../../sdc
```

### Remove LUN 0 on target and login again
I do the next:
1) Logout from target:
```$ sudo iscsiadm -m node --logout```
2) Remove LUN 0 on target
3) Login, show sessions and mounted disks by id again:
```
$ sudo -s iscsiadm --mode node --portal 192.168.13.116:3260 --targetname iqn.2010-06.com.nutanix:volumegroup-2-0a51e6cb-1666-4feb-809c-8d38110f17c2 --login
$ sudo iscsiadm -m session -P 3
iSCSI Transport Class version 2.0-870
version 2.0-873
Target: iqn.2010-06.com.nutanix:volumegroup-2-0a51e6cb-1666-4feb-809c-8d38110f17c2 (non-flash)
	Current Portal: 192.168.13.116:3260,1
	Persistent Portal: 192.168.13.116:3260,1
		**********
		Interface:
		**********
		Iface Name: default
		Iface Transport: tcp
		Iface Initiatorname: iqn.1993-08.org.debian:01:5bb3c3d5f39
		Iface IPaddress: 192.168.12.227
		Iface HWaddress: <empty>
		Iface Netdev: <empty>
		SID: 364
		iSCSI Connection State: LOGGED IN
		iSCSI Session State: LOGGED_IN
		Internal iscsid Session State: NO CHANGE
		*********
		Timeouts:
		*********
		Recovery Timeout: 120
		Target Reset Timeout: 30
		LUN Reset Timeout: 30
		Abort Timeout: 15
		*****
		CHAP:
		*****
		username: <empty>
		password: ********
		username_in: <empty>
		password_in: ********
		************************
		Negotiated iSCSI params:
		************************
		HeaderDigest: None
		DataDigest: None
		MaxRecvDataSegmentLength: 262144
		MaxXmitDataSegmentLength: 1048576
		FirstBurstLength: 262144
		MaxBurstLength: 16776192
		ImmediateData: Yes
		InitialR2T: Yes
		MaxOutstandingR2T: 1
		************************
		Attached SCSI devices:
		************************
		Host Number: 366	State: running

$ ll /dev/disk/by-id/
total 0
drwxr-xr-x 2 root root  60 жов 16 11:27 ./
drwxr-xr-x 5 root root 100 жов  4 16:32 ../
lrwxrwxrwx 1 root root   9 жов  4 16:32 ata-VMware_Virtual_IDE_CDROM_Drive_10000000000000000001 -> ../../sr0
```
As you see there is no any devices, but target really contains two LUN's: `20` and `40`

### Add LUN 0 on target and login again

Then when I add LUN with number `0` on target, then all works again and I see all my devices with my LUN's: `0, 20, 40`

### iscsiadm version
```
$ iscsiadm --version
iscsiadm version 2.0-873
```

**Comment by gonzoleeman** [comment]
> On Oct 16, 2019, at 8:46 AM, Chris Leech <notifications@github.com> wrote:
> 
> I believe that a SCSI target must respond to basic commands on LUN 0, even if there is no storage device attached as LUN 0, as that's how the REPORT_LUNS command works. I'd need to see a network trace of the session login and SCSI scanning (INQUIRY and REPORT_LUNS commands) to determine if the target or initiator is misbehaving in this case.
> 
> 

Yes, according to the SCSI standard, each target must respond to REPORT LUNS on LUN 0.

— 
Lee

**Comment by cleech** [comment]
I quickly tested this between a Linux initiator and target, and it worked as expected.  LU 0 responded to INQUIRY (with an indicator of no actual logical unit attached) and REPORT LUNS and higher numbered LUs functioned as expected.

I'm not familiar with Nutanix or their iSCSI target implementation.

**Comment by morden2k** [comment]
Ok. Thank you guys. This is really invalid issue and this specified by standard. I found this in SAM 5 (SCSI Architecture Model - 5) 4.7.4:

> All SCSI target devices shall support LUN 0 (i.e., 00000000 00000000h) or the REPORT LUNS well known logical unit. For SCSI target devices that support complex LUN structures (see 4.7.6) the LUN 0 or the REPORT LUNS well known logical unit shall be the logical unit that an application client addresses to determine information about the SCSI target device and the logical units contained within the SCSI target device.

So I really need some device with LUN 0 in any case.

**Comment by CST-hayata** [comment]
Sorry for responding to a closed issue, but I ran into this problem recently.  I investigate into it and find that my target did respond to LUN0 inquiry, but with a empty vendor/model name.

I then tried different kernel version. This issue exists with my original kernel 3.18, and disappears if I use linux kernel > 4.4.  After digging into the code, I found these line in drivers/scsi/scsi_devinfo.c:
```
if (memcmp(devinfo->vendor, vendor,
    min(max, strlen(devinfo->vendor))))
continue;

/* some other code */

if (memcmp(devinfo->model, model,
    min(max, strlen(devinfo->model))))
continue;
```
Basically this is matching the vendor/model sent by target with a list of device need to be specially treated.  With both string being empty, it matches the very first item in the list, making SCSI driver think this is that device in the list, and not sending a REPORT_LUNS as a result of that.
Later Linux kernel fix this issue by also checking the length of vendor/model string.
